Reconstruction/RecCalorimeter/src/components/CreateCaloCells.cpp

```diff
@@ -0,0 +1,217 @@
+  declareProperty("fieldValues", m_fieldValues);
+
+}
+
+CreateCaloCells::~CreateCaloCells()
+{
+}
+
+
+
+StatusCode CreateCaloCells::initialize() {
+
+  StatusCode sc = GaudiAlgorithm::initialize();
+  if (sc.isFailure()) return sc;
+
+  info() << "CreateCaloCells initialized" << endmsg;
+  info() << "doCellCalibration : " << m_doCellCalibration << endmsg;
+  info() << "addCellNoise      : " << m_addCellNoise << endmsg;
+  info() << "filterCellNoise   : " << m_filterCellNoise << endmsg;
+
+  //Initialization of various tools
+  //Merge hits
+  if (!m_mergeTool.retrieve()) {
+    error()<<"Unable to retrieve the merge hits to cells tool!!!"<<endmsg;
+    return StatusCode::FAILURE;
+  }
+
+  //Calibrate Geant4 energy to EM scale tool
+  if (m_doCellCalibration) {
+    if (!m_calibTool.retrieve()) {
+      error()<<"Unable to retrieve the calo cells calibration tool!!!"<<endmsg;
+      return StatusCode::FAILURE;
+    }
+  }
+
+  //Cell noise tool
+  if (m_addCellNoise) {
+    if (!m_noiseTool.retrieve()) {
+      error()<<"Unable to retrieve the calo cells noise tool!!!"<<endmsg;
+      return StatusCode::FAILURE;
+    }
+    //Prepare a vector of all cells in calorimeter with their cellID
+    StatusCode sc_prepareCells = PrepareEmptyCells(m_edmCellsVector);
+    if (sc_prepareCells.isFailure()) {
+      error()<<"Unable to create empty cells!!! Check the input for readout"<<endmsg;
+      return StatusCode::FAILURE;
+    }
+  }
+    
+  return sc;
+}
+
+StatusCode CreateCaloCells::execute() {
+   
+  //Get the input collection with Geant4 hits
+  const fcc::CaloHitCollection* hits = m_hits.get();
+  debug() << "Input Hit collection size: " << hits->size() << endmsg;
+
+  //Final vector of cells
+  std::vector<fcc::CaloHit*> edmFinalCellsVector;
+
+  //Merge hits with the same cellID
+  std::vector<fcc::CaloHit*> edmMergedHitsVector = m_mergeTool->MergeOneCollection(*hits);
+
+  //Calibrate Geant4 energy to EM scale tool
+  if (m_doCellCalibration) {
+    m_calibTool->Calibrate(edmMergedHitsVector);
+  }
+  
+ 
+  if (!m_addCellNoise) {
+    edmFinalCellsVector = edmMergedHitsVector;
+  }
+  //Create random noise hits for each cell, merge noise hits with signal hits, apply filter on cell energy if required
+  else {
+    m_noiseTool->CreateRandomCellNoise(m_edmCellsVector);
+    edmFinalCellsVector = m_mergeTool->MergeTwoVectors(edmMergedHitsVector, m_edmCellsVector);
+    if (m_filterCellNoise) {
+      m_noiseTool->FilterCellNoise(edmFinalCellsVector);
+    }
+  }
+
+  //Copy information from vector to CaloHitCollection
+  fcc::CaloHitCollection* edmCellsCollection = new fcc::CaloHitCollection();
+  for (auto ecells : edmFinalCellsVector) {
+    fcc::CaloHit newCell = edmCellsCollection->create();
+      newCell.Core().Energy = ecells->Core().Energy;     
+      newCell.Core().Time = ecells->Core().Time;
+      newCell.Core().Bits = ecells->Core().Bits;
+      newCell.Core().Cellid = ecells->Core().Cellid;
+  }
```

Comment:
I think at the moment you leak the content of `edmFinalCellsVector` and sometimes `edmMergedHitsVector` (in the case where you add noise), as the `CaloHit` stored in there are not reference counter. This could be a place to clean that up.

Ideally, this should be taken care of with passing around the edm collection in the end.